src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs

```diff
@@ -236,16 +236,17 @@ private static bool MatchesLookupInfo(TagHelperDescriptor descriptor, LookupInfo
                 return false;
             }
 
-            // We need to escape the TypePattern so we can choose to only allow specific regex.
-            var escaped = Regex.Escape(lookupInfo.TypePattern);
+            if (lookupInfo.TypePattern.EndsWith("*"))
+            {
+                if (lookupInfo.TypePattern.Length == 1)
+                {
+                    return true;
+                }
 
-            // We surround the escaped with ^ and $ in order ot ensure a regex match matches the entire
-            // string. We also replace any '*' or '?' characters with regex to match appropriate content.
-            // '*' matches 0 or more characters lazily and '?' matches 1 character.
-            var pattern = "^" + escaped.Replace(@"\?", ".").Replace(@"\*", ".*?") + "$";
-            var regex = new Regex(pattern, RegexOptions.Singleline, Constants.RegexMatchTimeout);
+                return descriptor.TypeName.StartsWith(lookupInfo.TypePattern.Substring(0, lookupInfo.TypePattern.Length - 1));
```

Comment:
Would inverting the check work: `lookupInfo.TypePattern.IndexOf(descriptor.TypeName, 0, lookupInfo.TypePattern - 1) == 0`